### PR TITLE
`date_part` support fractions of second

### LINKED ISSUE #3997: change` date_part` return types to f64
`date_part` currently  returns `i32` which might lose some information, e.g.

```bash
❯ select date_part('second', timestamp '2000-01-01T00:00:00.1');
+--------------------------------------------------------+
| datepart(Utf8("second"),Utf8("2000-01-01T00:00:00.1")) |
+--------------------------------------------------------+
| 0                                                      |
+--------------------------------------------------------+
1 row in set. Query took 0.000 seconds.
```

while postgresql returns double precision
```bash
willy=# select date_part('second', timestamp '2000-01-01T00:00:00.1');
 date_part 
-----------
       0.1
(1 row)
```

is it recommended to follow postgresql's return type?
if it is preferred, perhaps we could do a separate pr to deal with this

change it to double precision

**Describe the solution you'd like**

**Describe alternatives you've considered**

perhaps we could consider return decimals instead to align with extract to return decimal. postgres doc mentions this:
```
For historical reasons, the date_part function returns values of type double precision. 
This can result in a loss of precision in certain uses. Using extract is recommended instead.
```
https://www.postgresql.org/docs/current/functions-datetime.html

**Additional context**

### PR DESCRIPTION
**Pull request by comphead**
# Which issue does this PR close?

Closes #3997

# Rationale for this change
See #3997 

# What changes are included in this PR?
Extending `date_part` to support millis, micros, nanos

# Are these changes tested?
Yes

# Are there any user-facing changes?

No

No

### CODE REVIEW

**Comment by comphead** [comment]
@waitingkuo finally we can get back to #3997 resolution.
Please check the PR.

Now date_part supports second fraction, but its still not the same as in PSQL like 
```
willy=# select date_part('second', timestamp '2000-01-01T00:00:00.1');
 date_part 
-----------
       0.1
(1 row)
```

To do the same you need to run
```
select date_part('milliseconds', timestamp)
```

Let me know if this is ok

**Comment by waitingkuo** [comment]
hi @comphead  thank you

i tested the pr
```bash
❯ select date_part('second', timestamp '2000-01-01T00:00:09.123456');
+-------------------------------------------------------------+
| datepart(Utf8("second"),Utf8("2000-01-01T00:00:09.123456")) |
+-------------------------------------------------------------+
| 9                                                           |
+-------------------------------------------------------------+
1 row in set. Query took 0.002 seconds.
❯ select date_part('millisecond', timestamp '2000-01-01T00:00:09.123456');
+------------------------------------------------------------------+
| datepart(Utf8("millisecond"),Utf8("2000-01-01T00:00:09.123456")) |
+------------------------------------------------------------------+
| 123                                                              |
+------------------------------------------------------------------+
1 row in set. Query took 0.002 seconds.
❯ select date_part('microsecond', timestamp '2000-01-01T00:00:09.123456');
+------------------------------------------------------------------+
| datepart(Utf8("microsecond"),Utf8("2000-01-01T00:00:09.123456")) |
+------------------------------------------------------------------+
| 123456                                                           |
+------------------------------------------------------------------+
1 row in set. Query took 0.002 seconds.
```

the behavior is different than what postgresql has
```bash
willy=# select date_part('second', timestamp '2000-01-01T00:00:09.123456');
 date_part 
-----------
  9.123456
(1 row)

willy=# select date_part('millisecond', timestamp '2000-01-01T00:00:09.123456');
 date_part 
-----------
  9123.456
(1 row)

willy=# select date_part('microsecond', timestamp '2000-01-01T00:00:09.123456');
 date_part 
-----------
   9123456
(1 row)
```

i checked some other system,
spark seems to have the same behavior as postgresql
```bash
# this is spark
SELECT date_part('SECONDS', timestamp'2019-10-01 00:00:01.000001');
+----------------------------------------------------------+
|date_part(SECONDS, TIMESTAMP '2019-10-01 00:00:01.000001')|
+----------------------------------------------------------+
|                                                  1.000001|
+----------------------------------------------------------+
```

while mysql's is similar as this pr
```bash
# this is MYSQL
EXTRACT(SECOND FROM "2017-06-20 00:00:01.123456");
1
```

I originally purposed to output f64 instead of i32 since i'd like to follow postgresql's

@alamb  @tustvold  do you have any suggestion?

**Comment by comphead** [comment]
Thanks @waitingkuo for testing the PR. I think since we now having seconds fraction support so we can concat the value and be in sync with postgres date_part if this is a preferred way

**Comment by comphead** [comment]
@waitingkuo made in sync with postgres.
I have found extract nanosecond faces parser issue
```
DataFusion CLI v14.0.0
❯ select EXTRACT(nanosecond FROM to_timestamp('2020-09-08T12:00:12.12345678+00:00'));  🤔 Invalid statement: sql parser error: Expected date/time field, found: nanosecond
``` 

whereas nanosecond date_part works. I will create a ticket for this

File: datafusion/core/tests/sql/expr.rs [hunk 1]
```diff
@@ -1285,12 +1285,36 @@ async fn test_extract_date_part() -> Result<()> {
...
+        "12123.45678"
+    );
+    test_expression!(
+        "EXTRACT(microsecond FROM to_timestamp('2020-09-08T12:00:12.12345678+00:00'))",
+        "12123456.78"
+    );
+    // test_expression!(
+    //     "EXTRACT(nanosecond FROM to_timestamp('2020-09-08T12:00:12.12345678+00:00'))",
+    //     "1212345678"
+    // );
```
Comment (waitingkuo) [inline datafusion/core/tests/sql/expr.rs:1302]: i think we could uncomment this test case and expect the error
e.g. https://github.com/apache/arrow-datafusion/blob/740a4fa2c6ba4b85875a433bb86e5b00435a5969/datafusion/core/tests/sql/timestamp.rs#L1071-L1074

**Review by waitingkuo — APPROVED** [review]
hi @comphead thank you, this looks good to me.

**Comment by waitingkuo** [comment]
> @waitingkuo made in sync with postgres. I have found extract nanosecond faces parser issue
> 
> ```
> DataFusion CLI v14.0.0
> ❯ select EXTRACT(nanosecond FROM to_timestamp('2020-09-08T12:00:12.12345678+00:00'));  🤔 Invalid statement: sql parser error: Expected date/time field, found: nanosecond
> ```
> 
> whereas nanosecond date_part works. I will create a ticket for this

@comphead sqlparser doesn't support nanosecond for now.
https://github.com/sqlparser-rs/sqlparser-rs/blob/813f4a2eff8f091b643058ac1a46a4fbffd96806/src/parser.rs#L1187-L1221

**Comment by waitingkuo** [comment]
@comphead i created a ticket and submitted a pr in sqlparser sqlparser-rs/sqlparser-rs#748

it would be great if you could creat another ticket in datafusion so that we could fix it once the issue in sqlparser merged and new version released

**Comment by comphead** [comment]
> @comphead i created a ticket and submitted a pr in sqlparser [sqlparser-rs/sqlparser-rs#748](https://github.com/sqlparser-rs/sqlparser-rs/issues/748)
> 
> it would be great if you could creat another ticket in datafusion so that we could fix it once the issue in sqlparser merged and new version released

https://github.com/apache/arrow-datafusion/issues/4528
Please assign this to me.

**Comment by alamb** [comment]
Thanks @comphead  and @waitingkuo !

**Pull request merged**